src/duk_debugger.c

```diff
@@ -1639,6 +1690,37 @@ DUK_LOCAL void duk__debug_process_message(duk_hthread *thr) {
 	return;
 }
 
+/* Halt execution and enter a debugger message loop until execution is resumed
+ * by the client. PC for the current activation may be temporarily decremented
+ * so that the "current" instruction will be highlighted in the client.
+ */
+
+DUK_INTERNAL void duk_debug_halt_execution(duk_hthread *thr, duk_bool_t use_prev_pc) {
+	duk_activation *act;
+	duk_instr_t *old_pc;
+
+	DUK_HEAP_SET_PAUSED(thr->heap);
+
+	/* Decrement PC if that was requested, this requires a PC sync */
+	act = duk_hthread_get_current_activation(thr);
+	duk_hthread_sync_currpc(thr);
+	old_pc = act->curr_pc;
+	if (use_prev_pc && act->curr_pc > 0) {
+		act->curr_pc--;
+	}
+
+	/* Process debug messages until we are no longer paused */
+	thr->heap->dbg_processing = 1;
+	duk_debug_send_status(thr);
+	while (thr->heap->dbg_paused) {
+		DUK_ASSERT(thr->heap->dbg_processing);
+		duk_debug_process_messages(thr, 0 /* no_block */);
+	}
+	thr->heap->dbg_processing = 0;
+
+	act->curr_pc = old_pc;  /* restore PC */
```

Comment:
I addressed the style issues and error semantics in the latest commit.

I completely understand the thing about not looking up globals to enforce semantics, by the way--I had to do a similar thing with minisphere  a while back.  My engine is meant to be compatible with the old Sphere engine, which didn't have any constructors for built-in objects, and I ran into an issue where legacy games were clobbering the constructors, causing mysterious errors when something like `LoadImage()` was called.  I had to switch to putting the built-in prototypes in the global stash so that the engine could still construct builtins without having to look up the class name directly.